### ISSUE TITLE
Midi sequencing (feature request / question)

### ISSUE THREAD

**Issue by Gimmeapill** [harryhaaren/openAV-Luppp#98]
Creating a new feature request entry as I didn't want to hijack the Sequencer view thread.

I'm also interested in loop based Midi sequencing (ableton/seq24 like) but not so much by audio (hence the separate thread).
Playing live is in my experience a lot funnier if you add synths/generators and not rely only on pre-rendered audio loops - for this you need midi or OSC.

And as far as free software on linux/BSD goes there's still a gap:
Seq24 is pretty much dead and was anyway quite limited (display, number of tracks, midi controller mapping), Non-Sequencer is really too weird for daily use, and the Linux DAWs don't really like live looping. 
Luppp seems to have gotten the design part fine as an Ableton like loop triggering tool (congrats) - but it doesn't do midi and I was wondering how far away this is in the feature request list.

I think what would be needed as a bare minimum for a dedicated performance tool (without implementing a full fledged Midi editor/timeline/piano roll/tracker), is the ability to load individual midi files dynamically and play them in sync - maybe also save them in a session folder.

This way you could possibly save yourself the pain of reimplementing something that was already done by the studio tools out there, and focus only on the live performance part: play the file, but leave the creation duty to other tools - since everyone seem to have their favorite midi editor.
One file per slot - that could be reloded while playing. Not sure if that really helps, just an idea.

Some references:

A few live midi sequencing alternatives mentioned on Linux Musicians:
http://cvsweb-scrape.bzzt.net/viewtopic.php?f=1&t=13010
None seems ready for prime time though

A discussion about step sequencing on the Ardour forums:
https://community.ardour.org/node/6929

Existing proprietary solutions on Linux:
http://www.renoise.com/tools/live-dive
http://www.bitwig.com/en/home/recent-news.html

+ another suggestion regarding the OpenAV business model:
Wouldn't it make sense to add a "what's coming next and how much was donated for it" next to the donation buttons?
As a user, I like the concept of being able to vote with my wallet for a particular upcoming feature ;-)

Cheers (and sorry for the long post).

LX

**Comment by harryhaaren** [comment]
Hi Lx,

Thanks for your suggestions and feature request. The answer to the "MIDI seq" question is a bit complicated - but I'll do my best.

- Totally agree that MIDI sequencing / looping / is a core part of live-performance. The exact workflow that would be "ideal" on stage still isn't 100% clear in my head, but I have come to understand that although MIDI file playback would be useful (and fun) its not the long-term-goal here.

- An in-development version of Luppp actually has an initial implementation of MIDI looping (recording and playback, no save / loading files). The user-interface isn't ready, and the implementation isn't ideal either - but as proof of concept it works.

- The larger issue is that Luppp should be an LV2 plugin - this would allow it to run on the MOD platform[1], which is an amazing piece of tech that OpenAV is pretty excited about :)

- Right now as OpenAV I'm focusing on a particular use-case, a guitar/vocals setup, and integrating that with the MOD. Over time, more performance oriented features like MIDI looping etc will be added.

- Re OpenAV Release System: there's going to be some changes in that department. There's a negative side to the whole, which wasn't quite obvious to me when using it initially. I really do need to do a little video explaining this - thanks for the reminder. 

- I understand the "vote with the wallet" concept is nice, however its not sustainable at time of writing. I hope to change things up a little in terms of financial etc, so lets see.

- In short, Luppp probably won't get MIDI looping, the looping-project-as-LV2-plugin is probably of much more interest to you. Its currently not yet announced, however I will be demonstrating it a little at the Linux Audio Conference in a few weeks time. Stay tuned!

Cheers, -Harry

[1] http://portalmod.com

**Comment by Gimmeapill** [comment]
Hi Harry,

**Comment by Gimmeapill** [comment]
Hi Harry,

Thanks for the fast reply and sharing a bit of your roadmap.

Ok, so no midi looping in sight for Luppp, fair enough -  The "Looping as LV2" concept sound sensible on the long term, especially if you see it as an accompaniment tool to integrate it with the MOD or any LV2 host (as a dilettante guitar player I can also relate to that).

Regarding the workflow for playing live electronic music on a computer (in front of an audience or at home): I do have a few years of experience, and know pretty well what I need. Nothing particularly fancy, but still suprisingly hard to achieve with the current set of tools on Linux. 
Even without looking for a fully integrated environment a la Ableton, the pain point always end up being the sequencer. Otherwise I'm pretty happy with Ardour/Qtractor + existing softsynths and samplers.

But enough talk, I'll be looking forward your LAC presentation with great interest ;-)

Cheers,

LX

**Comment by harryhaaren** [comment]
If you have experience in Live-looping, then I'm all ears as to what exactly it is you'd like to see. I have an idea myself, but if you don't mind typing out the basics of what you need, then i'll certainly take it into account :)

Cheers, -Harry

**Comment by Gimmeapill** [comment]
I do, but I'm an old school tracker freak, so maybe not such a good example when it comes to modern sequencer design ;-)

But ok, here's how I do things: a Windows tracker (Jeskola Buzz) running on Wine - with a workaround for live playing called LiveJump (http://www.buzzmachines.com/machineinfo.php?id=859 - Full user manual there: http://cdn7.atwikiimg.com/_pub/buzzusers/backup/BuzzMachines_help_backup/BTDSys%20LiveJumpHACK.html).

LiveJump allows to define the boundaries of each loop section in the main sequencer and binds it to a keyboard key or midi event (can be a note or cc).
When triggered, the corresponding loop section gets scheduled and start at the end of the current loop - this affects all tracks at once (I believe this is the same as changing scene in Ableton).
There is not  much of a GUI except for a digital countdown to the end of the currently playing loop section - it does look pretty cool though.
This setup is not as flexible as Ableton, but good enough for what I have to do. I've been using it for the last 11 years, played ~50-60 live events, so it's battle tested.
Another advantage to the minimalistic GUI: you hardly ever need to look at the screen when playing live - this kills the performance and should be avoided.

But as said earlier, the Luppp interface looks pretty fine to me: it follows the Ableton paradigm, which is probably the fastest way to lure in some user base - it would just needs midi/sequencing capabilities to fit my needs.

On the other hand, if you go the LV2 route (a distinct project as I understand), I'd also be happy with a minimalistic step sequencer that could be inserted per track in a DAW.
This should be quite close from the workflow you get with a hardware based liveset and grooveboxes (each machine comes with it's own sequencer, patterns can be changed independently), and provide live capabilities to exiting DAWs without reinventing the wheel. Would probably also play better with the MOD as I guess this thing doesn't have unlimited power.

Basic set of features that come to mind:
Sync with the host (of course)
1 track per plugin instance, not sure if multitrack would make much sense here.
Pattern size: 8-16-32-64-128 steps
number of patterns: let's say 128 (maybe organize them in banks?)
midi file import/save (preferably outside of the DAW session)
pattern # triggerable via midi/OSC/computer Keyboard (either instant change or scheduled at the end of the currently playing pattern).
UI: not sure if much more than the LV2 default controls would be even needed apart for the Midi file loading dialog. Unless you plan there a fully fledged sequencer.

For reference, the windows software "AudioMulch" features small sequencer modules in that vein (albeit much more limited): http://www.audiomulch.com/info/what-is-audiomulch.

Again just ideas...I hope some of this makes sense...

Cheers,

LX

**Comment by Gimmeapill** [comment]
Actually, it looks like falkTX has also something in mind when it comes to sequencing within his Carla LV2 host: http://www.linuxmusicians.com/viewtopic.php?f=1&t=12968&start=15

Need to check out the status of that one as it seems quite close to what I described earlier...

**Comment by harryhaaren** [comment]
Hi @Gimmeapill, thanks for these suggestions - they've been read (and now re-read), I agree that there are some nice ideas around there... lets see :)

I'm currently cleaning up the Luppp issue tracker - so I'm going to mark this as closed. If you have more ideas, or other comments to make, then feel free to post them here anyway! Thanks again, -Harry

**Comment by Gimmeapill** [comment]
Looking forward to your new creations ;-)
